src/__tests__/bag_browser.js

```diff
@@ -0,0 +1,163 @@
+import React from "react";
+import BagBrowser from "../bag_browser";
+import {render, fireEvent, cleanup} from "react-testing-library";
+
+let doc;
+
+const searchField = () => doc.baseElement.getElementsByTagName("input").item(0);
+const fireSearch = query => fireEvent.change(searchField(), { target: { value: query } });
+
+afterEach(() => {
+  cleanup();
+  fetch.resetMocks();
+});
+
+describe('<BagBrowser/>', () => {
+  beforeEach(() => {
+    doc = render(
+      <BagBrowser/>
+    );
+  });
+
+  it("creates an empty text input", () => {
+    expect(searchField().getAttribute("type")).toEqual("text");
+    expect(searchField().getAttribute("value")).toEqual("");
+  });
+
+  it("has some text telling the user to enter an identifier", () => {
+    expect(doc.baseElement.innerHTML).toMatch(
+      /Enter an identifier to view a package/);
+  });
+
+  describe("when searching for a 404ing package", () => {
+    beforeEach(() => {
+      fetch.mockResponseOnce('Not found', { status: 404 });
+      fireSearch("wrong_id")
+    });
+
+    it('should call fetch("v1/bags/wrong_id")', () => {
+      expect(fetch.mock.calls.length).toEqual(1);
+      expect(fetch.mock.calls[0][0]).toEqual("v1/bags/wrong_id");
+    });
+
+    it("displays an error message with 404 in it", () => {
+      expect(doc.baseElement.innerHTML).toMatch(/could not retrieve/i);
+      expect(doc.baseElement.innerHTML).toMatch(/wrong_id/i);
+      expect(doc.baseElement.innerHTML).toMatch(/404/i);
+    });
+  });
+
+  describe("if the api is broken and returning garbage json", () => {
+    beforeEach(() => {
+      fetch.mockResponseOnce("{{{{{{{{{{");
+      fireSearch("uh_oh_its_bad")
+    });
+
+    it('should call fetch("v1/bags/uh_oh_its_bad")', () => {
+      expect(fetch.mock.calls.length).toEqual(1);
+      expect(fetch.mock.calls[0][0]).toEqual("v1/bags/uh_oh_its_bad");
+    });
+
+    it("displays an error message with 'json' in it", () => {
+      expect(doc.baseElement.innerHTML).toMatch(/could not retrieve/i);
+      expect(doc.baseElement.innerHTML).toMatch(/uh_oh_its_bad/i);
+      expect(doc.baseElement.innerHTML).toMatch(/json/i);
+    });
+  });
```

Comment:
Does this construct work? We normally see the expectations set on the mocks and then the trigger... That is, I would expect to see `fireSearch` as the last statement of these two examples.

Reply:
I don't understand the question. Before each test, it mocks `fetch` and then fires an event that will cause `fetch` to be called. Then, in each test, we are testing the state after fetch has been called, and the component updated accordingly. Does that answer your question?

Reply:
Right -- the event being fired seems part of the example, rather than the setup. How I usually see this is like:

```
before {
  // make mock(s), configuring responses
}

it {
  expect(mock).toReceive(...)
  thingUnderTest.someOperation()
  // maybe an explicit verify step, but not usually with the details
}
```

The expect-then-trigger pattern shows up more in the BDD style tests. Not a big difference, really; just observing a stylistic thing and that I guess these mocks allow looking back at call history.

Reply:
I guess my thinking was "I want this to happen before more than one test, so it should go in beforeEach" rather than about any particular paradigm. Should I change it? Right now I'm like, cool, no duplicate code, but I know there are more important things than keeping DRY. Is this one of them?